arch/arm/core/cortex_m/mpu/arm_core_mpu.c

```diff
@@ -10,105 +10,294 @@
+		u32_t num_partitions = mem_domain->num_partitions;
+		struct k_mem_partition partition;
+		int i;
 
-void _arch_mem_domain_configure(struct k_thread *thread)
-{
-	configure_mpu_mem_domain(thread);
+		LOG_DBG("configure domain: %p", mem_domain);
+
+		for (i = 0; i < CONFIG_MAX_DOMAIN_PARTITIONS; i++) {
+			partition = mem_domain->partitions[i];
+			if (partition.size == 0) {
+				/* Zero size indicates a non-existing
+				 * memory partition.
+				 */
+				continue;
+			}
+			LOG_DBG("set region 0x%x 0x%x",
+				partition.start, partition.size);
+			dynamic_regions[region_num].start = partition.start;
+			dynamic_regions[region_num].size = partition.size;
+			dynamic_regions[region_num].attr = partition.attr;
+
+			region_num++;
+			num_partitions--;
+			if (num_partitions == 0) {
+				break;
+			}
+		}
+	}
+	/* Thread user stack */
+	LOG_DBG("configure user thread %p's context", thread);
+	if (thread->arch.priv_stack_start) {
+		u32_t base = (u32_t)thread->stack_obj;
+		u32_t size = thread->stack_info.size;
+#if !defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
+		/* In user-mode the thread stack will include the (optional)
+		 * guard area. For MPUs with arbitrary base address and limit
+		 * it is essential to include this size increase, to avoid
+		 * MPU faults.
+		 */
+		size += thread->stack_info.start - base;
+#endif
+		dynamic_regions[region_num].start = base;
+		dynamic_regions[region_num].size = size;
+		dynamic_regions[region_num].attr = K_MEM_PARTITION_P_RW_U_RW;
+
+		region_num++;
+	}
+#endif /* CONFIG_USERSPACE */
+
+#if defined(CONFIG_MPU_STACK_GUARD)
+	/* Privileged stack guard */
+#if defined(CONFIG_USERSPACE)
+	u32_t guard_start = thread->arch.priv_stack_start ?
+	    (u32_t)thread->arch.priv_stack_start :
+	    (u32_t)thread->stack_obj;
+#else
+	u32_t guard_start = thread->stack_info.start;
+#endif
+	dynamic_regions[region_num].start = guard_start;
+	dynamic_regions[region_num].size = MPU_GUARD_ALIGN_AND_SIZE;
+	dynamic_regions[region_num].attr = K_MEM_PARTITION_P_RO_U_NA;
+
+	region_num++;
+#endif /* CONFIG_MPU_STACK_GUARD */
+
+	/* Configure the dynamic MPU regions */
+	arm_core_mpu_configure_dynamic_mpu_regions(dynamic_regions,
+		region_num);
 }
 
+#if defined(CONFIG_USERSPACE)
+
+/**
+ * @brief Get the maximum number of partitions for a memory domain
+ *        that is supported by the MPU hardware.
+ */
 int _arch_mem_domain_max_partitions_get(void)
 {
 	return arm_core_mpu_get_max_domain_partition_regions();
 }
 
-/*
- * Reset MPU region for a single memory partition
+/**
+ * @brief Configure the memory domain of the thread.
  */
-void _arch_mem_domain_partition_remove(struct k_mem_domain *domain,
-				       u32_t  partition_id)
+void _arch_mem_domain_configure(struct k_thread *thread)
 {
-	ARG_UNUSED(domain);
+	/* Request to configure memory domain for a thread.
+	 * This triggers re-programming of the entire dynamic
+	 * memory map.
+	 */
+	_arch_configure_dynamic_mpu_regions(thread);
```

Comment:
You are right @andyross, this is a bit confusing, and, _slightly_ problematic, but I believe we can live with that.

In brief:
- `_arch_mem_domain_configure()` is an existing API from `kernel/mem_domain.h`. As such, I want to preserve it as is, in this re-work of the ARM MPU.
- `_arch_configure_dynamic_mpu_regions ()` is the new, single, universal API function for re-programming the Memory map at run-time (user-mode enter and context-switch). As such, `_arch_configure_dynamic_mpu_regions ()` also programs the (app) memory domains.

Now: app memory domains may _also_ be programmed at run-time at arbitrary places (i.e. any privileged thread, or IRQ is allowed do it). I just felt that I could **_reuse_** the existing function of programming the dynamic memory map, instead of adding a new ARM Core MPU function that programs only the memory partitions.

The comment:
```
/*
 * This triggers re-programming of the entire dynamic
 * memory map.
 */

indicates my true intentions.

>Can't you just use the old function name universally?

This will be done in the future, actually: I am going to use the `_arch_configure_dynamic_mpu_regions()` even inside the kernel/mem_domain.c.
But not in this PR :)
```